substrate/serai/client/src/in_instructions.rs

```diff
@@ -0,0 +1,53 @@
+use scale::Decode;
+
+use serai_runtime::{
+  support::traits::PalletInfo as PalletInfoTrait, PalletInfo, in_instructions, InInstructions,
+  Runtime,
+};
+
+pub use in_instructions_primitives as primitives;
+
+use crate::{
+  primitives::{Coin, BlockNumber},
+  Serai, SeraiError,
+};
+
+const PALLET: &str = "InInstructions";
+
+pub type InInstructionsEvent = in_instructions::Event<Runtime>;
+
+impl Serai {
+  pub async fn get_batch_events(
+    &self,
+    block: [u8; 32],
+  ) -> Result<Vec<InInstructionsEvent>, SeraiError> {
+    let mut res = vec![];
+    for event in
+      self.0.events().at(Some(block.into())).await.map_err(|_| SeraiError::RpcError)?.iter()
+    {
+      let event = event.map_err(|_| SeraiError::InvalidRuntime)?;
+      if PalletInfo::index::<InInstructions>().unwrap() == usize::from(event.pallet_index()) {
+        let mut with_variant: &[u8] =
+          &[[event.variant_index()].as_ref(), event.field_bytes()].concat();
+        let event =
+          InInstructionsEvent::decode(&mut with_variant).map_err(|_| SeraiError::InvalidRuntime)?;
+        if matches!(event, InInstructionsEvent::Batch { .. }) {
+          res.push(event);
+        }
+      }
+    }
+    Ok(res)
+  }
+
+  pub async fn get_coin_block_number(
+    &self,
+    coin: Coin,
+    block: [u8; 32],
+  ) -> Result<BlockNumber, SeraiError> {
+    Ok(self.storage(PALLET, "BlockNumbers", Some(coin), block).await?.unwrap_or(BlockNumber(0)))
```

Comment:
To be clear, we are publishing the block number for each coin to substrate now?

Reply:
We need to in order to coordinate Monero properly.

If Serai doesn't have the inputs available, the first block at which they are available can be used without question. If Serai does, then the question is "is it block X, my current block, or block X -1, what everyone else has due to latency?" Placing the height on chain solves that.

It also provides a heartbeat, usable to detect desyncs, which helps with #6.